backend/services/implementations/statisticsService.ts

```diff
@@ -0,0 +1,147 @@
+
+  constructor(
+    testSessionService: ITestSessionService,
+    testService: ITestService,
+    userService: IUserService,
+    schoolService: ISchoolService,
+  ) {
+    this.testSessionService = testSessionService;
+    this.testService = testService;
+    this.userService = userService;
+    this.schoolService = schoolService;
+  }
+
+  async getAverageScoresBySchool(
+    testId: string,
+  ): Promise<Array<StatisticsResponseDTO>> {
+    let statisticsDTOs: Array<StatisticsResponseDTO> = [];
+
+    try {
+      const statistics: AggregateTestSessionDTO[] = await MgTestSession.aggregate(
+        [
+          // Stage 1: Filter Test Session documents by test id
+          {
+            $match: { test: Types.ObjectId(testId) },
+          },
+          // Stage 2: Group remaining documents by school id and add aggregate fields
+          {
+            $unwind: "$results",
+          },
+          {
+            $group: {
+              _id: "$school",
+              averageScore: { $avg: "$results.score" },
+              resultsPerTestArr: {
+                $push: {
+                  $map: {
+                    input: "$results.breakdown",
+                    as: "result",
+                    in: { $cond: ["$$result", 1, 0] },
+                  },
+                },
+              },
+            },
+          },
+          // Stage 3: Sort documents by averageScore in descending order
+          {
+            $sort: { averageScore: -1 },
+          },
+        ],
+      );
+
+      /* eslint-disable no-param-reassign */
+      statistics.forEach((statistic: AggregateTestSessionDTO) => {
+        statistic.averageScoresByQuestions = this.computeAverageScorePerQuestion(
+          statistic.resultsPerTestArr,
+        );
+      });
+
+      statisticsDTOs = await this.mapStatisticsToStatisticsDTOs(
+        testId,
+        statistics,
+      );
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to get average scores by school. Reason = ${getErrorMessage(
+          error,
+        )}`,
+      );
+      throw error;
+    }
+
+    return statisticsDTOs;
+  }
+
+  private computeAverageScorePerQuestion(
+    resultsPerTestArr: number[][],
+  ): number[] {
+    const averageResultsPerQuestion: number[] = [];
+    const numOfQuestions = resultsPerTestArr[0].length;
+    const numOfResults = resultsPerTestArr.length;
+
+    for (let i = 0; i < numOfQuestions; i += 1) {
+      let scoresSum = 0;
+
+      resultsPerTestArr.forEach((test: number[]) => {
+        scoresSum += test[i];
+      });
+
+      const averageScore = +((scoresSum * 100) / numOfResults).toFixed(2);
+      averageResultsPerQuestion.push(averageScore);
+    }
+
+    return averageResultsPerQuestion;
+  }
+
+  private async mapStatisticsToStatisticsDTOs(
+    testId: string,
+    statistics: AggregateTestSessionDTO[],
+  ): Promise<Array<StatisticsResponseDTO>> {
+    const testSessionDtos: Array<StatisticsResponseDTO> = await Promise.all(
```

Comment:
Not sure if I'm just brainfreezing, but why is this called `testSessionDtos`?

Reply:
LOL forgot to rename - thanks for catching that!